lib/streamlit/hashing.py

```diff
@@ -408,3 +404,13 @@ def _code_to_bytes(self, code, context):
                     self._update(h, name)
 
         return h.digest()
+
+    @staticmethod
+    def _get_main_script_directory():
+        """Get the directory of the main script.
+        """
+        import __main__
+        import os
+
+        main_path = __main__.__file__
```

Comment:
Does this point to the right file? In some toy tests on my computer, it seems to point to the equivalent of `cli.py`

Here's what I did:

```python
# main_test.py

with open('main_subtest.py') as script_file:
    script = script_file.read()

exec(script, {})
```

```python
# main_subtest.py

import streamlit as st
import __main__
st.write(__main__.__file__)
```

Expected: should see "main_subtest.py" in Streamlit app
Actual: saw "main_test.py" in Streamlit app

Reply:
I tested this in `streamlit`... In the case below, it shows the correct path. Let me try to repro. I would rather use `_report.script_path` but we need to pass it down to the hasher, which I could do, but I need to change the interface.

<img width="756" alt="Screen Shot 2019-10-08 at 11 42 32 AM" src="https://user-images.githubusercontent.com/6821233/66423647-f1216100-e9c0-11e9-8b9a-a48abe404bb7.png">

Reply:
Interesting:
https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/ScriptRunner.py#L293

We do this, before running: 
```
sys.modules["__main__"] = module
```

Reply:
Actually, happens here:```
module.__dict__["__file__"] = self._report.script_path
```

Reply:
Which is what we are looking for, right?
I think if we like this solution we need at the very least a test.

Reply:
You're right. This code is fine, then!

Can you add a comment before line 415 explaining why this is fine?

Reply:
Will add the comment.